fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java

```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
             int perPage,
             int currentPage
     ) {
+        /**
+         * Works around the lack of a way to get at the underlying Collector in TimeLimitingCollector
+         */
+        class AccessibleTimeLimitingCollector extends TimeLimitingCollector {
+            private TopScoreDocCollector wrappedCollector;
+
+            /**
+             * @param collector the wrapped collector
+             * @param clock the clock used to implement the timeout
+             * @param ticksAllowed the amount of time allowed for the search
+             */
+            AccessibleTimeLimitingCollector(TopScoreDocCollector collector, Counter clock, long ticksAllowed) {
+                super(collector, clock, ticksAllowed);
+                wrappedCollector = collector;
+            }
+
+            /**
+             * This is the method TimeLimitingCollector was missing.
+             * @return the wrapped collector
+             */
+            public TopScoreDocCollector getWrappedCollector() {
+                return wrappedCollector;
+            }
+        }
+
+        /**
+         * This parallelizes the lucene search - each thread using a time limited collector.
+         */
+        final CollectorManager<AccessibleTimeLimitingCollector, TopDocs> manager
+                = new CollectorManager<AccessibleTimeLimitingCollector, TopDocs>() {
+
+            @Override
+            public AccessibleTimeLimitingCollector newCollector() throws IOException {
+                return new AccessibleTimeLimitingCollector(
+                        TopScoreDocCollector.create(perPage, lastEntry), Counter.newCounter(false), searchTimeout
+                );
+            }
+
+            @Override
+            public TopDocs reduce(Collection<AccessibleTimeLimitingCollector> collectors) throws IOException {
+                final TopDocs[] docs = collectors.stream()
+                        .map((collector) -> collector.getWrappedCollector().topDocs())
+                        .collect(Collectors.toList())
+                        .toArray(new TopDocs[collectors.size()]);
+
+                return TopDocs.merge(perPage, docs);
+            }
+        };
```

Comment:
It would be great to pull this anonymous class (and the helper wrapper class) out into it's own stand-alone class, for a number of reasons.

1. I think the capability is generally useful, especially if another part of the application needs to deal with Lucene in a different way. Having time-bounded searches is generally a good idea
2. Having these classes defined in-line in the method will incur more performance overhead, since a new class is generated every method call, rather than in a more static manner that will happen once per class-load
3. The classes are very hard to unit-test when they are embedded in the method. Making them top-level makes them much easier to test on their own as a small, focused unit with little environment / setup overhead
4. The `LuceneSearchProvider` is already rather large. Adding more responsibilities to it adds more complexity, but having this responsibility (ie. time-bounding TopDoc searches) encapsulated into it's own top-level class keeps the complexity better in-check

It probably makes sense to have a single top-level `TimeLimitingTopDocsCollectorManager` class, with the `AccessibleTimeLimitingCollector` as an inner-class, since it's pretty tightly associated with time-limiting top docs.

I would also leave the `searchTimeout` in the `LuceneSearchProvider`, and pass it to the `TimeLimitingTopDocsCollectorManager` as a constructor param, since we've generally found it better to keep the interactions with `SystemConfig` as high in the stack as we can to reduce coupling and keep responsibilities small for pieces low in the stack.